modules/deployment-scripts/create-agw-kv-certificate/main.bicep

```diff
@@ -0,0 +1,229 @@
+  scope: akv
+  properties: {
+    roleDefinitionId: resourceId('Microsoft.Authorization/roleDefinitions', roleDefId)
+    principalId: useExistingManagedIdentity ? existingDepScriptId.properties.principalId : newDepScriptId.properties.principalId
+    principalType: 'ServicePrincipal'
+  }
+}]
+
+resource rbacAppGw 'Microsoft.Authorization/roleAssignments@2020-08-01-preview'  =  [for roleDefId in rbacRolesNeededOnAppGw: if (!empty(agwName) && agwCertType != 'none') {
+  name: guid(agw.id, roleDefId, useExistingManagedIdentity ? existingDepScriptId.id : newDepScriptId.id)
+  scope: agw
+  properties: {
+    roleDefinitionId: resourceId('Microsoft.Authorization/roleDefinitions', roleDefId)
+    principalId: useExistingManagedIdentity ? existingDepScriptId.properties.principalId : newDepScriptId.properties.principalId
+    principalType: 'ServicePrincipal'
+  }
+}]
+
+var managedIdentityOperator = resourceId('Microsoft.Authorization/roleDefinitions', 'f1a07417-d97a-45cb-824c-7a7467783830')
+@description('Managed identity requires "Managed Identity Operator" permission over the user assigned identity of the Application Gateway.')
+resource agwIdRbac 'Microsoft.Authorization/roleAssignments@2020-08-01-preview' =  if (!empty(agwName) && agwCertType != 'none') {
+  scope: agwId
+  name: guid(agwId.id, managedIdentityOperator, useExistingManagedIdentity ? existingDepScriptId.id : newDepScriptId.id)
+  properties: {
+    roleDefinitionId: managedIdentityOperator
+    principalId: useExistingManagedIdentity ? existingDepScriptId.properties.principalId : newDepScriptId.properties.principalId
+    principalType: 'ServicePrincipal'
+  }
+}
+
+@batchSize(1)
+resource createImportCert 'Microsoft.Resources/deploymentScripts@2020-10-01' = [for certName in certNames: {
+  name: 'AKV-Cert-${akv.name}-${replace(replace(certName,':',''),'/','-')}'
+  location: location
+  identity: {
+    type: 'UserAssigned'
+    userAssignedIdentities: {
+      '${useExistingManagedIdentity ? existingDepScriptId.id : newDepScriptId.id}': {}
+    }
+  }
+  kind: 'AzureCLI'
+  dependsOn: [
+    rbacKv
+    rbacAppGw
+  ]
+  properties: {
+    forceUpdateTag: forceUpdateTag
+    azCliVersion: azCliVersion
+    timeout: timeout
+    retentionInterval: retention
+    environmentVariables: [
+      {
+        name: 'RG'
+        value: resourceGroup().name
+      }
+      {
+        name: 'akvName'
+        value: akvName
+      }
+      {
+        name: 'agwName'
+        value: agwName
+      }
+      {
+        name: 'agwCertType'
+        value: agwCertType
+      }
+      {
+        name: 'certName'
+        value: certName
+      }
+      {
+        name: 'initialDelay'
+        value: initialScriptDelay
+      }
+      {
+        name: 'certificateWait'
+        value: agwCertImportDelay
+      }
+    ]
+    scriptContent: '''
+      #!/bin/bash
+      set -e
+
+      echo "Waiting on Identity RBAC replication ($initialDelay)"
+      sleep $initialDelay
```

Comment:
we can be more deterministic by doing a GET on the roleAssignment and breaking if we only see 200's

Reply:
Aren't there nuances if the SP will be querying its own role assignments? 
Just running a;

```azurecli
az role assignment list --assignee $spClientId
```

returns an empty array.

I'm not sure if adding more RBAC just so the SP can check it's own RBAC is going to be worth it.

If you have a sample of this working well @bmoore-msft i'd be keen to look

Reply:
We do something similar here: https://github.com/Azure/azure-quickstart-templates/blob/master/test/ci-scripts/Create-ResourceGroupAssignment.ps1 - pass in the ```rbacKv.id``` and just do a few GETs on it.  But I think I agree with you on the extra perms the MSI would need to query the graph, it's not as elegant.

The other thing we could do is try/catch on the ```az keyvault certificate create```.  If it's failing with 4xx, then we know we can retry within reason.  Chances are, by the time the deploymentScript resource spins up, the role has replicated...  So you wouldn't need many retries IMO.

Reply:
Try/Catch seems the best option here.
I'll leave the initialDelay parameter in there as a cushion though.

Bash doesn't really do try/catch, so i've looped it with a maximum retry attempt.
Tested, and works nice. I even specifically denied the RBAC assignment and it fails gracefully.